File: tools/pack.js
const Jimp = require('jimp')

const settings = {
    basePath: './ftp2/YOKO_Park/',
    outPath: './textures/',
    scale: 0.5,
    animTextureSize: 2048,
    map: {
        path: 'Park/YokoPark_Background.png',
        outName: 'map',
        textureSize: 1024,
        quality: 80
    },
    animations: [
        {
            path: 'Живые деокрации - Ambient_animations/Водопад/_YokoSite_Waterfall_Sprite.png',
            outName: 'waterfall',
            width: 950,
            height: 950,
            x: 4000,
            y: 750,
            numFrames: 100
        },
        {
            path: 'Живые деокрации - Ambient_animations/Блики на воде1/_YokoSite_Water_Sprite.png',
            outName: 'water1',
            width: 1340,
            height: 580,
            x: 3200,
            y: 1480,
            numFrames: 31
        },
        {
            path: 'Живые деокрации - Ambient_animations/Блики на воде2/_YokoSite_Water_2_Sprite.png',
            outName: 'water2',
            width: 1600,
            height: 600,
            x: 1100,
            y: 2260,
            numFrames: 31
        },
        {
            path: 'Живые деокрации - Ambient_animations/Домик тети Маши/_YokoSite_Light_Sprite.png',
            outName: 'lights',
            width: 420,
            height: 300,
            x: 3840,
            y: 2220,
            numFrames: 15,
            textureSize: 1024
        },
        {
            path: 'Волшебные превращения - FXs/Песочница - Замок/_YokoSite_Castle_Sprite.png',
            outName: 'castle',
            width: 1060,
            height: 910,
            x: 1560,
            y: 1270,
            numFrames: 30
        },
        {
            path: 'Волшебные превращения - FXs/Коряга - Пиратский корабль/_YokoSite_Ship_Sprite.png',
            outName: 'ship',
            width: 1120,
            height: 830,
            x: 1020,
            y: 1980,
            numFrames: 87
        },
        {
            path: 'Волшебные превращения - FXs/Вороны - Пингвины/_YokoSite_Crows_Sprite.png',
            outName: 'penguins',
            width: 580,
            height: 600,
            x: 2840,
            y: 1940,
            numFrames: 45,
            textureSize: 1024
        },
        {
            path: 'Волшебные превращения - FXs/Статуя Йоко - Йоко/_YokoSite_Statue_Sprite.png',
            outName: 'statue',
            width: 1040,
            height: 960,
            x: 2420,
            y: 1400,
            numFrames: 30
        },
        {
            path: 'Волшебные превращения - FXs/Елка - Ракета/_YokoSite_Rocket_Sprite.png',
            outName: 'rocket',
            width: 490,
            height: 610,
            x: 2860,
            y: 950,
            numFrames: 34
        }
    ]
}

Jimp.read(settings.basePath + settings.map.path).then(function(mapImage) {
    const mapScaled = mapImage.clone().scale(settings.scale)

    console.log('Map loaded.')

    const texSize = settings.map.textureSize

    const mapWidth = mapScaled.bitmap.width
    const mapHeight = mapScaled.bitmap.height

    const texNumX = Math.ceil(mapWidth / texSize)
    const texNumY = Math.ceil(mapHeight / texSize)

    for (let i = 0; i < texNumY; i++) {
        for (let j = 0; j < texNumX; j++) {
            new Jimp(texSize, texSize, function(err, outImage) {
                const x = j * texSize
                const y = i * texSize
                const w = x + texSize > mapWidth ? mapWidth - x : texSize
                const h = y + texSize > mapHeight ? mapHeight - y : texSize

                const outPath = settings.outPath + settings.map.outName + (i * texNumX + j) + '.png'
                //outImage.blit(mapScaled, 0, 0, x, y, w, h).quality(settings.map.quality).write(outPath)
                outImage.blit(mapScaled, 0, 0, x, y, w, h).deflateLevel(9).write(outPath, function() {

                })

                console.log(`Saved map tile ${x} ${y} to ${outPath}`)
            })
        }
    }

    for (let anim of settings.animations) {
        Jimp.read(settings.basePath + anim.path).then(function(animImage) {
            //animImage = animImage.scale(settings.scale)

            console.log(`Animation '${anim.outName}' loaded.`)

            const texSize = anim.textureSize || settings.animTextureSize
            const animWidth = animImage.bitmap.width
            const animHeight = animImage.bitmap.height
            const frameWidth = anim.width * settings.scale
            const frameHeight = anim.height * settings.scale
            const framesPerX = Math.floor(texSize / frameWidth)
            const framesPerY = Math.floor(texSize / frameHeight)
            const framesPerTexture = framesPerX * framesPerY
            //const numFrames = animWidth / frameWidth
            const numFrames = anim.numFrames

            new Jimp(anim.width, anim.height, function(err, tempImage) {
                //tempImage.blit(mapImage, 0, 0, anim.x, anim.y, anim.width, anim.height)

                let currentFrame = 0
                let currentTexture = 0

                function doAnimation() {
                    new Jimp(texSize, texSize, function(err, outImage) {

                        function finalizeAnimation() {
                            const outPath = settings.outPath + anim.outName + currentTexture + '.png'
                            outImage.deflateLevel(9).write(outPath, function() {

                            })

                            console.log(`Saved animation '${anim.outName}' texture ${currentTexture} to ${outPath}`)

                            currentTexture++
                        }

                        while (true) {
                            if (currentFrame >= numFrames) {
                                finalizeAnimation()
                                return
                            }

                            const currentTextureFrame = currentFrame - framesPerTexture * currentTexture
                            const x = currentTextureFrame % framesPerX
                            const y = Math.floor(currentTextureFrame / framesPerX)

                            if (y >= framesPerY) {
                                finalizeAnimation()
                                doAnimation()
                                return
                            }

                            new Jimp(anim.width, anim.height, function(err, frameImage) {
                                frameImage.blit(animImage, 0, 0, currentFrame * anim.width, 0, anim.width, anim.height)

                                const compImage = tempImage.clone()
                                //compImage.composite(frameImage, 0, 0)
                                compImage.blit(frameImage, 0, 0)
                                compImage.scale(settings.scale)

                                outImage.blit(compImage, x * frameWidth, y * frameHeight)
                            })

                            currentFrame++
                        }
                    })
                }

                doAnimation()
            })
        })
    }
})

File: tools/pack.py
import math
import json
from collections import OrderedDict
from PIL import Image

settings = {
    'jsonInFile': './animations.json',
    'jsonOutFile': '../frontend/public/textures.json',
    'basePath': '../ftp/YOKO_Park/',
    'outPath': '../frontend/public/images/textures/',
    'divisor': 3,
    'animTextureSize': 2048,
    'map': {
        'path': 'Park/YokoPark_Background_NEW.png',
        'outName': 'map',
        'textureSize': 512,
        'quality': 90
    }
}

with open(settings['jsonInFile'], ) as jsonFile:
    animations = json.load(jsonFile)

basePath = settings['basePath']
mapSettings = settings['map']

mapImage = Image.open(basePath + mapSettings['path'])
mapSize = mapImage.size

outData = {
    'scale': 1.0 / settings['divisor'],
    'map': {
        'name': mapSettings['outName'],
        'textureSize': mapSettings['textureSize'],
        'width': mapSize[0],
        'height': mapSize[1]
    }
}

texSize = mapSettings['textureSize']
divisor = settings['divisor']
mapWidth = mapSize[0] / divisor
mapHeight = mapSize[1] / divisor

mapScaled = mapImage.resize((mapWidth, mapHeight), Image.LANCZOS)
texNumX = int(math.ceil(float(mapWidth) / texSize))
texNumY = int(math.ceil(float(mapHeight) / texSize))

for i in range(texNumY):
    for j in range(texNumX):
        x = j * texSize
        y = i * texSize
        w = x + (mapWidth - x) if x + texSize > mapWidth else texSize
        h = y + (mapHeight - y) if y + texSize > mapHeight else texSize

        texImage = Image.new('RGB', (texSize, texSize))
        texImage.paste(mapScaled.crop((x, y, x + w, y + h)))

        # outPath = settings['outPath'] + \
        #    mapSettings['outName'] + str(i * texNumX + j) + '.jpg'
        #texImage.save(outPath, quality=mapSettings['quality'])

        outPath = settings['outPath'] + \
            mapSettings['outName'] + str(i * texNumX + j) + '.png'
        texImage.save(outPath)
        print "Saved map texture %d, %d" % (j, i)


def finalizeAnimation(outImage, anim, currentTexture):
    outPath = settings['outPath'] + \
        anim['outName'] + str(currentTexture) + '.png'
    outImage.save(outPath)
    print "Saved animation '%s' texture %d" % (anim['outName'], currentTexture)

outData['animations'] = OrderedDict()

for category, animList in animations.items():
    for anim in animList:
        animImage = Image.open(basePath + anim['path'])

        texSize = settings['animTextureSize']
        if 'textureSize' in anim:
            texSize = anim['textureSize']

        animWidth = animImage.size[0]
        animHeight = animImage.size[1]

        animDivWidth = anim['width'] - anim['width'] % divisor
        animDivHeight = anim['height'] - anim['height'] % divisor

        frameWidth = animDivWidth / divisor
        frameHeight = animDivHeight / divisor
        framesPerX = texSize / frameWidth
        framesPerY = texSize / frameHeight
        framesPerTexture = framesPerX * framesPerY

        numFrames = anim[
            'numFrames'] if 'numFrames' in anim else animWidth / anim['width']

        animData = {
            'width': animDivWidth,
            'height': animDivHeight,
            'numFrames': numFrames,
            'textureSize': texSize,
            'type': category
        }

        if 'x' in anim:
            animData['x'] = anim['x']

        if 'y' in anim:
            animData['y'] = anim['y']

        if 'speed' in anim:
            animData['speed'] = anim['speed']

        outData['animations'][anim['outName']] = animData

        currentFrame = 0
        currentTexture = 0

        outImage = Image.new('RGBA', (texSize, texSize))

        while True:
            currentTextureFrame = currentFrame - framesPerTexture * currentTexture
            x = currentTextureFrame % framesPerX
            y = currentTextureFrame / framesPerX

            if y >= framesPerY:
                finalizeAnimation(outImage, anim, currentTexture)
                outImage = Image.new('RGBA', (texSize, texSize))
                currentTexture += 1
                continue

            rect = (currentFrame * anim['width'], 0, currentFrame *
                    anim['width'] + animDivWidth, animDivHeight)
            frameImage = animImage.crop(rect).resize(
                (frameWidth, frameHeight), Image.LANCZOS)
            outImage.paste(frameImage, (x * frameWidth, y * frameHeight))

            currentFrame += 1

            if currentFrame >= numFrames:
                finalizeAnimation(outImage, anim, currentTexture)
                break

    with open(settings['jsonOutFile'], 'w') as jsonFile:
        json.dump(outData, jsonFile)
